src/Control/Carrier/Telemetry.hs

```diff
@@ -0,0 +1,147 @@
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Control.Carrier.Telemetry (
+  TelemetryC,
+  IgnoreTelemetryC,
+  withTelemetry,
+  withoutTelemetry,
+  bracket',
+) where
+
+import Control.Concurrent.STM (atomically, putTMVar, tryReadTMVar)
+import Control.Concurrent.STM.TBMQueue (tryWriteTBMQueue)
+import Control.Effect.Exception (Exception (fromException), catch, mask, throwIO)
+import Control.Effect.Lift (Lift, sendIO)
+import Control.Effect.Telemetry (Telemetry (..))
+import Control.Monad (void, when)
+import Control.Monad.IO.Class (MonadIO (..))
+import Data.Maybe (isNothing)
+
+import Control.Carrier.Telemetry.Types (
+  TelemetryCtx (..),
+  TelemetryTimeSpent (TelemetryTimeSpent),
+  TimedLogRecord (TimedLogRecord),
+ )
+import Control.Carrier.Telemetry.Utils (
+  mkTelemetryCtx,
+  mkTelemetryRecord,
+ )
+import Control.Exception qualified as Exc
+import Data.Time.Clock (diffUTCTime, getCurrentTime, nominalDiffTimeToSeconds)
+import Data.Tracing.Instrument (incCount)
+import System.Exit (ExitCode (ExitSuccess))
+
+import Control.Algebra (Algebra (alg), Has, type (:+:) (L, R))
+import Control.Carrier.Reader (ReaderC, ask, runReader)
+import Control.Carrier.Telemetry.Sink.Common (emitTelemetry)
+
+newtype TelemetryC m a = TelemetryC {runTelemetryC :: ReaderC TelemetryCtx m a}
+  deriving (Applicative, Functor, Monad, MonadFail, MonadIO)
+
+instance (Algebra sig m, MonadIO m, Has (Lift IO) sig m) => Algebra (Telemetry :+: sig) (TelemetryC m) where
+  alg hdl sig ctx = case sig of
+    L op -> do
+      telCtx <- TelemetryC (ask @(TelemetryCtx))
+      case op of
+        SetTelemetrySink sink -> do
+          maybeSink <- sendIO $ atomically $ tryReadTMVar (telSink telCtx)
+          when (isNothing maybeSink) $
+            sendIO $ atomically $ putTMVar (telSink telCtx) sink
+          ctx <$ pure ()
```

Comment:
Few notes:
- factor out `sendIO . atomically`, I called it `sendSTM`.
- `telCtx` can be used as a wildcard to greatly reduce needed parens
- `STM` ops should be grouped together when possible, and then run with `sendSTM` after.
- `putTMVar` can _theoretically_ block forever if `SetTelemetrySink` is called twice in quick succession before being used before (I know we don't do that, but caution is important in concurrent code).  We can use `void $ tryPutTMVar` to prevent blocking.  Doing that means that we don't need to check the `TMVar` first, which simplifies `TrackConfig` and `SetTelemetrySink`.
- [later]: Ideally, we would run everything here in `sendSTM $ case op of`, but the `getCurrentTime` function is preventing this.  I wonder if we can push those actions higher and use this effect as an `STM` interpreter.  We can do that later.
- move `sendSTM` as close to the top of the function stack as you can.  example: 
```haskell
SetTelemetrySink sink -> sendSTM (ctx <$ tryPutTMVar telSink sink) -- good
SetTelemetrySink sink -> ctx <$ (sendSTM $ tryPutTMVar telSink sink) -- bad, harder to refactor without switching them first
```

```suggestion

sendSTM :: Has (Lift IO) sig m => STM a -> m a
sendSTM = sendIO . atomically

instance (Algebra sig m, MonadIO m, Has (Lift IO) sig m) => Algebra (Telemetry :+: sig) (TelemetryC m) where
  alg hdl sig ctx = case sig of
    L op -> do
      TelemtryCtx {..} <- TelemetryC $ ask @TelemetryCtx
      case op of
        SetTelemetrySink sink -> sendSTM (ctx <$ tryPutTMVar telSink sink)
```

A lot of the comments for this carrier are based on this suggestion, so I would probably take them all-or-nothing.